File: sdk/python/src/graft/modules/neopixel.py
"""GRAFT NeoPixel Module"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graft.device import GraftDevice

from graft.core.constants import CMD


class GraftNeoPixel:
    def __init__(self, device: GraftDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def begin(self, pin: int, count: int, pixel_type: int = 0) -> None:
        payload = bytes([self._mod_id, 0x01, pin,
                         count & 0xFF, (count >> 8) & 0xFF, pixel_type])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def set_pixel(self, index: int, r: int, g: int, b: int, w: int = 0) -> None:
        payload = bytes([self._mod_id, 0x02,
                         index & 0xFF, (index >> 8) & 0xFF, r, g, b])
        if w:
            payload += bytes([w])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def set_range(self, start: int, count: int, r: int, g: int, b: int) -> None:
        payload = bytes([self._mod_id, 0x03,
                         start & 0xFF, (start >> 8) & 0xFF,
                         count & 0xFF, (count >> 8) & 0xFF, r, g, b])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def fill(self, r: int, g: int, b: int, w: int = 0) -> None:
        payload = bytes([self._mod_id, 0x04, r, g, b])
        if w:
            payload += bytes([w])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def show(self) -> None:
        payload = bytes([self._mod_id, 0x05])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def set_brightness(self, brightness: int) -> None:
        payload = bytes([self._mod_id, 0x06, brightness & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

File: sdk/js/src/modules/neopixel.ts
/**
 * GRAFT NeoPixel Module — Typed JS wrapper
 */

import type { GraftDevice } from '../device.js'

export class GraftNeoPixel {
  private _device: ReturnType<GraftDevice['module']>

  constructor(device: GraftDevice) {
    this._device = device.module('neopixel')
  }

  /** Initialize the NeoPixel strip. type: 0 = NEO_GRB+800KHz (default). */
  async begin(pin: number, count: number, type = 0): Promise<void> {
    const buf = new Uint8Array(4)
    buf[0] = pin
    buf[1] = count & 0xFF; buf[2] = (count >> 8) & 0xFF
    buf[3] = type
    await this._device.cmd(0x01, buf)
  }

  /** Set a single pixel color. */
  async setPixel(index: number, r: number, g: number, b: number, w = 0): Promise<void> {
    const buf = new Uint8Array(5 + (w ? 1 : 0))
    buf[0] = index & 0xFF; buf[1] = (index >> 8) & 0xFF
    buf[2] = r; buf[3] = g; buf[4] = b
    if (w) buf[5] = w
    await this._device.cmd(0x02, buf)
  }

  /** Set a range of pixels to the same color. */
  async setRange(start: number, count: number, r: number, g: number, b: number): Promise<void> {
    const buf = new Uint8Array(7)
    buf[0] = start & 0xFF; buf[1] = (start >> 8) & 0xFF
    buf[2] = count & 0xFF; buf[3] = (count >> 8) & 0xFF
    buf[4] = r; buf[5] = g; buf[6] = b
    await this._device.cmd(0x03, buf)
  }

  /** Fill all pixels with a color. */
  async fill(r: number, g: number, b: number, w = 0): Promise<void> {
    const buf = new Uint8Array(3 + (w ? 1 : 0))
    buf[0] = r; buf[1] = g; buf[2] = b
    if (w) buf[3] = w
    await this._device.cmd(0x04, buf)
  }

  /** Flush pixel buffer to the strip. */
  async show(): Promise<void> {
    await this._device.cmd(0x05)
  }

  /** Set global brightness (0-255). */
  async setBrightness(brightness: number): Promise<void> {
    await this._device.cmd(0x06, new Uint8Array([brightness & 0xFF]))
  }
}
